aliyun/src/main/java/org/apache/iceberg/aliyun/oss/OSSOutputStream.java

```diff
@@ -0,0 +1,164 @@
+import org.slf4j.LoggerFactory;
+
+public class OSSOutputStream extends PositionOutputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(OSSOutputStream.class);
+  private final StackTraceElement[] createStack;
+
+  private final OSS client;
+  private final OSSURI uri;
+
+  private final File currentStagingFile;
+  private final OutputStream stream;
+  private long pos = 0;
+  private boolean closed = false;
+
+  OSSOutputStream(OSS client, OSSURI uri, AliyunProperties aliyunProperties) {
+    this.client = client;
+    this.uri = uri;
+    this.createStack = Thread.currentThread().getStackTrace();
+
+    this.currentStagingFile = newStagingFile(aliyunProperties.ossStagingDirectory());
+    this.stream = newStream(currentStagingFile);
+  }
+
+  private static File newStagingFile(String ossStagingDirectory) {
+    try {
+      File stagingFile = File.createTempFile("oss-file-io-", ".tmp", new File(ossStagingDirectory));
+      stagingFile.deleteOnExit();
+      return stagingFile;
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  private static OutputStream newStream(File currentStagingFile) {
+    try {
+      return new BufferedOutputStream(new FileOutputStream(currentStagingFile));
+    } catch (FileNotFoundException e) {
+      throw new NotFoundException(e, "Failed to create file: %s", currentStagingFile);
+    }
+  }
+
+  private static InputStream uncheckedInputStream(File file) {
+    try {
+      return new FileInputStream(file);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void flush() throws IOException {
+    Preconditions.checkState(!closed, "Already closed.");
+    stream.flush();
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    Preconditions.checkState(!closed, "Already closed.");
+    stream.write(b);
+    pos += 1;
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    Preconditions.checkState(!closed, "Already closed.");
+    stream.write(b, off, len);
+    pos += len;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (closed) {
+      return;
+    }
+
+    super.close();
+    closed = true;
+
+    try {
+      stream.close();
+      completeUploads();
+    } finally {
+      cleanUpStagingFiles();
+    }
+  }
+
+  private void completeUploads() {
+    long contentLength = currentStagingFile.length();
+    if (contentLength == 0) {
+      LOG.debug("Skipping empty upload to OSS");
+      return;
+    }
+
+    LOG.debug("Uploading {} staged bytes to OSS", contentLength);
+    InputStream contentStream = uncheckedInputStream(currentStagingFile);
+    ObjectMetadata metadata = new ObjectMetadata();
+    metadata.setContentLength(contentLength);
+
+    PutObjectRequest request = new PutObjectRequest(uri.bucket(), uri.key(), contentStream, metadata);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: When objects are uploaded to Amazon S3 by using streams (either through an AmazonS3 client or TransferManager), you might encounter network connectivity or timeout issues.

**Fix**: The most reliable way to avoid a ResetException is to provide data by using a File or FileInputStream, which the AWS SDK for Java can handle without being constrained by mark and reset limits. If the input stream isn&#39;t a FileInputStream but does support mark and reset, you can set the mark limit by using the [setReadLimit](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/RequestClientOptions.html#setReadLimit-int-) method of [RequestClientOptions.](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/RequestClientOptions.html) Its default value is 128 KB. Setting the read limit value to one byte greater than the size of stream will reliably avoid a ResetException.

[Learn more](https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/best-practices.html)